libs/@local/hql/cst/src/symbol.rs

```diff
@@ -0,0 +1,292 @@
+use core::{fmt, fmt::Display};
+
+use smol_str::SmolStr;
+use unicode_ident::{is_xid_continue, is_xid_start};
+use winnow::{
+    combinator::{delimited, dispatch, empty, fail, opt, peek},
+    error::ParserError,
+    stream::{AsChar, Compare, Stream, StreamIsPartial},
+    token::{any, one_of, take_while},
+    PResult, Parser,
+};
+
+// TODO: in the future we might want to use the bump arena here as well.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Symbol(SmolStr);
+
+impl Display for Symbol {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        Display::fmt(&self.0, f)
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub(crate) enum ParseRestriction {
+    None,
+    RustOnly,
+}
+
+impl ParseRestriction {
+    const fn is_rust(self) -> bool {
+        matches!(self, Self::None | Self::RustOnly)
+    }
+
+    const fn is_operators(self) -> bool {
+        matches!(self, Self::None)
+    }
+}
+
+impl Default for ParseRestriction {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+/// Implementation of Symbol parsing
+///
+/// # Syntax
+///
+/// ```abnf
+/// symbol = regular / ignore / operator / operatorSafe
+///
+/// regular = XID_START *XID_CONTINUE
+/// ignore = "_" *XID_CONTINUE
+///
+/// operator = "+" / "-" / "*" / "/" / "|" / "&" / "^" / "==" / "!=" / ">" / ">=" / "<" / "<="
+/// operatorSafe = "`" operators "`"
+/// ```
+pub(crate) fn parse_symbol<Input, Error>(
+    restriction: ParseRestriction,
+) -> impl Parser<Input, Symbol, Error>
+where
+    Input: StreamIsPartial
+        + Stream<Token: AsChar + Clone, Slice: AsRef<str>>
+        + Compare<char>
+        + for<'a> Compare<&'a str>,
+    Error: ParserError<Input>,
+{
+    move |input: &mut Input| {
+        dispatch! {peek(any).map(|token: Input::Token| token.as_char());
```

Comment:
I don't know the `dispatch!` macro but it looks like a match. The first two match arms are identical, is that correct?

Reply:
They are not, the difference is

```
char if ... && is_xid_start(char)
```

vs

```
'_' if ...
```

`_` is not part of XID_Start (so like Rust) we need to special case it a bit